docs/design/llvm_sparse_runtime.md

```diff
@@ -0,0 +1,458 @@
+                                              +------------+
+                                              |     5      |
+                                              +------------+
+                                              |     6      |
+                                              +------------+
+                                              |     7      |
+                                              +------------+
+```
+
+The activation/deacctivation process is no different from that of `pointer`. We can trace through [`Dynamic_append`](https://github.com/taichi-dev/taichi/blob/0f4fb9c662e6e3ffacc26e7373258d8d0414423b/taichi/runtime/llvm/node_dynamic.h#L61-L87) to better understand the layout.
+
+```cpp
+i32 Dynamic_append(Ptr meta_, Ptr node_, i32 data) {
+  auto meta = (DynamicMeta *)(meta_);
+  auto node = (DynamicNode *)(node_);
+  auto chunk_size = meta->chunk_size;
+  // 1
+  auto i = atomic_add_i32(&node->n, 1);
+  // 2
+  int chunk_start = 0;
+  auto p_chunk_ptr = &node->ptr;
+  while (true) {
+    // 3
+    if (*p_chunk_ptr == nullptr) {
+      locked_task(Ptr(&node->lock), [&] {
+        if (*p_chunk_ptr == nullptr) {
+          auto rt = meta->context->runtime;
+          auto alloc = rt->node_allocators[meta->snode_id];
+          *p_chunk_ptr = alloc->allocate();
+        }
+      });
+    }
+    // 4
+    if (i < chunk_start + chunk_size) {
+      // 4-1
+      *(i32 *)(*p_chunk_ptr + sizeof(Ptr) +
+               (i - chunk_start) * meta->element_size) = data;
+      break;
+    }
+    // 4-2
+    p_chunk_ptr = (Ptr *)(*p_chunk_ptr);
+    chunk_start += chunk_size;
+  }
+  return i;
+}
+```
+
+1. Uses the current length `n` as the index (`i`) to store `data`.
+2. `chunk_strat` tracks the starting index of a given chunk, and always starts at `0` . `p_chunk_ptr` is initialized to the pointer to the first chunk.
+3. Inside the `while` loop, checks if the given chunk slot is empty first, and allocates a new chunk if so.
+4. Compares if the determined index `i` falls within the current chunk.
+   1. If so, stores `data` into the corresponding slot in this chunk. Note that the first `sizeof(Ptr)` bytes are skiped: they are reserved to store the address of the next chunk.
+   2. Otherwise, jumps to the next chunk.
+
+# Runtime
+
+Runtime for the LLVM backends is in https://github.com/taichi-dev/taichi/blob/master/taichi/runtime/llvm/runtime.cpp. Note that this file is *NOT* linked into Taichi's CPP core library. Instead, it is compiled into a LLVM byte code file (`.bc`). Upon starting Taichi, the `.bc` file is loaded back into the memory, de-serialized into an `llvm::Module`, and linked together with the JIT compiled Taichi kernels. This design has the advantage that the runtime code can be written once and shared between CPU and CUDA backends. In addition, the sparse runtime can be implemented in a language with enough abstraction (i.e., C++ instead of raw LLVM IR).
```

Comment:
```suggestion
Runtime for the LLVM backends is in https://github.com/taichi-dev/taichi/blob/master/taichi/runtime/llvm/runtime.cpp. Note that this file is *NOT* linked into Taichi's core C++ library. Instead, it is compiled into a LLVM byte code file (`.bc`). Upon starting Taichi, the `.bc` file is loaded back into the memory, de-serialized into an `llvm::Module`, and linked together with the JIT compiled Taichi kernels. This design has the advantage that the runtime code can be written once and shared between backends using LLVM, such as CPU and CUDA. In addition, the sparse runtime can be implemented in a language with enough abstraction (i.e., C++ instead of raw LLVM IR).
```